pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,182 @@
+	ExchangeKind       = "fanout"
+	QueueDurability    = true
+	QueueDelete        = false
+	QueueExclusivity   = false
+	QueueWait          = false
+	ConsumerTag        = "mainflux-consumer"
+	Mandatory          = false
+	Immediate          = false
+)
+
+var (
+	errAlreadySubscribed = errors.New("already subscribed to topic")
+	errNotSubscribed     = errors.New("not subscribed")
+	errEmptyTopic        = errors.New("empty topic")
+)
+
+var _ messaging.PubSub = (*pubsub)(nil)
+
+// PubSub wraps messaging Publisher exposing
+// Close() method for RabbitMQ connection.
+type PubSub interface {
+	messaging.PubSub
+	Close()
+}
+
+type pubsub struct {
+	conn          *amqp.Connection
+	logger        log.Logger
+	queue         amqp.Queue
+	channel       *amqp.Channel
+	subscriptions map[string]bool
+	done          chan bool
+}
+
+// NewPubSub returns RabbitMQ message publisher/subscriber.
+func NewPubSub(url, queueName string, logger log.Logger) (PubSub, error) {
+	endpoint := fmt.Sprintf("amqp://%s", url)
+	conn, err := amqp.Dial(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, err
+	}
+	queue, err := ch.QueueDeclare(queueName, QueueDurability, QueueDelete, QueueExclusivity, QueueWait, nil)
+	if err != nil {
+		return nil, err
+	}
+	ret := &pubsub{
+		conn:          conn,
+		queue:         queue,
+		channel:       ch,
+		logger:        logger,
+		subscriptions: make(map[string]bool),
+		done:          make(chan bool),
+	}
+	return ret, nil
+}
+
+func (ps *pubsub) Publish(topic string, msg messaging.Message) error {
+	if topic == "" {
+		return errEmptyTopic
+	}
+	data, err := proto.Marshal(&msg)
+	if err != nil {
+		return err
+	}
+	subject := fmt.Sprintf("%s.%s.%s", Exchange, ChansPrefix, topic)
+	if err := ps.channel.ExchangeDeclare(subject, ExchangeKind, true, false, false, false, nil); err != nil {
+		return err
+	}
+	err = ps.channel.Publish(
+		subject,
+		RoutingKey,
+		Mandatory,
+		Immediate,
+		amqp.Publishing{
+			Headers:     amqp.Table{},
+			ContentType: "text/plain",
+			Priority:    2,
+			AppId:       "mainflux",
+			Body:        []byte(data),
+		})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ps *pubsub) Subscribe(topic string, handler messaging.MessageHandler) error {
+	if topic == "" {
+		return errEmptyTopic
+	}
+	if _, ok := ps.subscriptions[topic]; ok {
+		return errAlreadySubscribed
+	}
+
+	subject := fmt.Sprintf("%s.%s.%s", Exchange, ChansPrefix, topic)
+
+	if err := ps.channel.ExchangeDeclare(subject, ExchangeKind, true, false, false, false, nil); err != nil {
+		return err
+	}
+
+	if err := ps.channel.QueueBind(ps.queue.Name, RoutingKey, subject, false, nil); err != nil {
+		return err
+	}
+	msgs, err := ps.channel.Consume(ps.queue.Name, "", true, false, false, false, nil)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		ps.handle(msgs, handler)
+		wg.Done()
+	}()
+
+	ps.subscriptions[topic] = true
```

Comment:
Should this map be protected for concurrent read/write?

Reply:
Usually by using a mutex.

Reply:
Done